### DISCUSSION TITLE
accumulate_grad_batches seems not working

### DISCUSSION THREAD

**Discussion by amejri** [Lightning-AI/pytorch-lightning#12166] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hi guys,

I'm trying to train a model with accumulating several batches for gradient. For that, I specified it in the trainer as follow :
```
    trainer = pl.Trainer(gpus=gpus, max_epochs=nb_epochs, distributed_backend="ddp", precision=16, accumulate_grad_batches=32)
```
but it seems like nothing changes. and the training evolution is the same as without adding accumulate_grad_batches parameter.
Do you have any idea how to solve it ?

With kind regards,

**Comment by akihironitta** [comment]
@amejri Seems like you're using an old version of PyTorch Lightning (<1.5.0 because the trainer flag `distributed_backend` was removed then). Would you mind updating its version to the stable one with `pip install pytorch-lightning -U` and trying it again? If the issue still persists, would it be possible for you to reproduce it with the [BoringModel](https://github.com/Lightning-AI/lightning/tree/9fc4ff3278ac360f2671549415f6b11689a263ed/examples/pl_bug_report)?

**Comment by rohitgr7** [comment]
hey @amejri !

> but it seems like nothing changes. and the training evolution is the same as without adding accumulate_grad_batches parameter.

can you share more on this? how are you evaluating this conclusion?

**Reply by cristi-zz** [reply]
Hi! I can add a new "datapoint" here.
1) Link to BoringModel is broken, leads to 404 on github.
2) I am doing Self Supervised Learning, this tutorial: https://docs.lightly.ai/tutorials/package/tutorial_moco_memory_bank.html . No multi-GPU.

I trained with BS=512 (deep red curve) then with BS=16 (blueish-magenta) and then, BS=16 AND accumulate_grad_batches=32 (Orange)

I am stopping the train right now. For me, the flat loss curve shows that the BS=16+accumulate_grad is not doing what it should. 
Did not poke into the lightning code. 

```
pytorch-lightning         1.7.7                    pypi_0    pypi
torch                     1.12.1                   pypi_0    pypi
torchmetrics              0.9.3                    pypi_0    pypi
torchvision               0.13.1                   pypi_0    pypi

```


![image](https://user-images.githubusercontent.com/4200067/192540676-d8121d7d-4b45-47e9-be56-853f3d86c9e6.png)

**LE**

I restarted with BS=512 AND accumulate_grad to 1. Light blue line, traces the "original" BS=512 training loss.
![image](https://user-images.githubusercontent.com/4200067/192543153-bfa201c6-085c-4778-98f6-112b9f330cbd.png)

**Reply by rjbruin** [reply]
I am also seeing different training behavior with and without `accumulate_grad_batches`. I have a training setting where I can run two models with the same command to identical loss training curves (using `pl.seed_everything` and PyTorch deterministic mode), but when I divide the batch size by four and use `accumulate_grad_batches=4` the loss training curves are very different.

**Reply by ivanstepanovftw** [reply]
Maybe BatchNorm is the cause?
